lib/TPP/Conversion/ConvertVectorToXsmm/ConvertTranspose.cpp

```diff
@@ -0,0 +1,405 @@
+        mlir::utils::getStaticStrides(transposeOp->getResult(0));
+    if (failed(stridesOnOutput) || stridesOnOutput->back() != 1)
+      return failure(transposeOp);
+    auto unaryInfo = xsmm::utils::getVectorUnaryInfo(
+        dyn_cast<MemRefType>(input->getOperand(0).getType()),
+        dyn_cast<MemRefType>(output->getOperand(1).getType()),
+        dyn_cast<VectorType>(input->getResult(0).getType()),
+        dyn_cast<VectorType>(output->getOperand(0).getType()),
+        xsmm::UnaryFlags::NONE);
+    if (failed(unaryInfo)) {
+      return failure(transposeOp);
+    }
+  } else {
+    if (!xsmm::utils::isTwoDTransposeOp(
+            dyn_cast<mlir::vector::TransposeOp>(transposeOp))) {
+      return failure(transposeOp);
+    }
+    transposeOp->dump();
+    auto unaryInfo = xsmm::utils::getVectorUnaryInfo(
+        dyn_cast<MemRefType>(input->getOperand(0).getType()),
+        dyn_cast<MemRefType>(output->getOperand(1).getType()),
+        dyn_cast<VectorType>(input->getResult(0).getType()),
+        dyn_cast<VectorType>(output->getOperand(0).getType()),
+        xsmm::UnaryFlags::NONE);
+    if (failed(unaryInfo)) {
+      return failure(transposeOp);
+    }
+  }
+  return success(transposeOp);
+}
+
+void registerNativeTransposeRewrite(RewritePatternSet &patterns) {
+  patterns.getPDLPatterns().registerRewriteFunction("BuildTranspose",
+                                                    buildTransposeOp);
+  patterns.getPDLPatterns().registerConstraintFunction("ValidateTranspose",
+                                                       validateTransposeOpImpl);
+}
+
+namespace mlir {
+namespace tpp {
+
+struct ConvertTranspose
+    : public PassWrapper<ConvertTranspose, OperationPass<>> {
+  MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(ConvertTranspose)
+
+  StringRef getArgument() const final { return "insert-transpose-pass"; }
+
+  StringRef getDescription() const final {
+    return "Insert transpose before gemms/brgemms and convert to XSMM "
+           "functionality";
+  }
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<pdl::PDLDialect, pdl_interp::PDLInterpDialect,
+                    mlir::vector::VectorDialect, func::FuncDialect,
+                    memref::MemRefDialect, LLVM::LLVMDialect, BuiltinDialect>();
+  }
+
+  LogicalResult initialize(MLIRContext *ctx) override {
+    // Build the pattern set within the `initialize` to avoid recompiling
+    // PDL patterns during each `runOnOperation` invocation.
+    RewritePatternSet patternList(ctx);
+    registerNativeTransposeRewrite(patternList);
+    populateGeneratedPDLLPatterns(patternList);
+    patterns = std::move(patternList);
+    return success();
+  }
+
+  void runOnOperation() final {
+    PatternRewriter rewriter(&getContext());
+    // Enable conversion for linalg.generic to XSMM Brgemm if possible.
+    auto res =
+        getOperation()->walk([&](mlir::vector::ContractionOp contractOp) {
```

Comment:
We want to rewrite transpose, why are we walking over contraction ops?

Reply:
Like i said earlier, I thought I'd have the code to insert transposes when gemms cannot be performed in this patch itself. If we want to keep this in the brgemm/gemm lowering patch, I'll move this to the next patch.